### PR TITLE
refactor(metrics): emit application metrics through Temporal SDK pipeline

### PR DESCRIPTION
**Pull request by solidDoWant**
## Summary

- Application metrics now flow through `activity.GetMetricsHandler` and `workflow.GetMetricsHandler` instead of registering Prometheus collectors directly. Histograms with custom bucket boundaries use `contribtally.ScopeFromHandler(...)` per the contrib/tally docs. The dedicated `RecordRunMetrics` and `RecordInvalid` activities, the `Recorder` type, and `MediaWorkflowConfig.MetricsRegisterer` are gone — each producing activity emits its own observations inline.
- End-to-end workflow latency is dropped from the application set; `temporal_workflow_endtoend_latency_seconds` (tagged `workflow_type`) covers it. Per-step activity latency is also available natively via `temporal_activity_execution_latency_seconds` (tagged `activity_type`). `media_workflow_transcode_duration_seconds` is kept as an application metric so transcode runtime can be sliced by codec / hardware acceleration / crop / HC tags — context the SDK metric does not carry.
- Replaces `contribtally.NewPrometheusNamingScope` with a local `promNamingScope` that auto-suffixes Counter (`_total`) and Timer (`_seconds`) like the upstream scope but leaves Histogram names untouched, so app histograms can use `_bytes` (or any non-time suffix) without being mangled. Embeds `tally.Scope` so only Counter/Timer/Tagged/SubScope need overriding.
- Bucket boundaries use `tally.MustMakeExponentialValueBuckets` / `MustMakeExponentialDurationBuckets`: file sizes 10MB → ~328GB (16 buckets, factor 2), durations 1m → ~8.5h (10 buckets, factor 2).

### Metric inventory changes

| Removed                                         | Replacement                                                                          |
| ----------------------------------------------- | ------------------------------------------------------------------------------------ |
| `media_workflow_total_duration_seconds`         | `temporal_workflow_endtoend_latency_seconds{workflow_type=\"MediaWorkflow\"}`        |

Application metric types in the new shape: histograms for `source_duration_seconds`, `source_file_size_bytes`, `destination_file_size_bytes`, `transcode_duration_seconds`; gauges for `audio_track_count` / `subtitle_track_count`; counters for `invalid_files_total`, `artwork_fetch_skipped_total`, `metrics_errors_total`. Every app metric also picks up the SDK-injected `namespace` / `task_queue` / `activity_type` / `workflow_type` tags automatically.

### Why some changes the way they are

- **High-cardinality tags still empty-string-padded for movies.** Tally's prom reporter caches metrics on `(name, sortedTagKeys)`. Two emissions of the same metric name with different tag-key sets attempt two Prometheus registrations, the second hits Prometheus's \"same fully-qualified name with different label names\" rejection, and the reporter silently replaces the second with a `noopMetric`. Omitting `series_title` / `season_number` / `episode_number` for movies would silently drop one media-type variant the first time both run on the same worker. Empty-string padding is the only path that doesn't depend on which media type registered first.
- **`tally.DefaultBuckets` not used.** It's a sentinel (\"use the reporter's default\"), not a configurable global, and one global default cannot fit both file sizes and durations.

### Docs

`docs/metrics.md` is rewritten for an operator audience: dropped the \"How metrics are emitted\" implementation section, the \"Emitted by\" column, and references to internal Go functions. Kept endpoint config, `METRICS_SCRAPE_WAIT_TIMEOUT` semantics + the K8s `terminationGracePeriodSeconds` interaction, the per-metric reference, the per-metric tag table, the SDK metrics that operators should rely on for workflow timing, and the cardinality warning on HC mode.

## Test plan

- [x] `make fmt` / `make vet` / `make lint` clean
- [x] `make test` — full unit suite passes with `-race`
- [x] New activity-level metric tests (`activities_test.go`) verify Probe emits the invalid-files counter for invalid input and the source-duration histogram + audio/subtitle gauges for valid media via `tally.NewTestScope`
- [x] `make test-integration` — not run; requires Docker daemon, which is unavailable in the dev container (\`docker\` not in PATH, no socket mounted). Needs to be run on a host with Docker before merge.
- [ ] After merge, scrape `/metrics` on a worker with both a movie and a show workflow run to confirm the new series + tag sets are landing as expected (both with and without `METRICS_HIGH_CARDINALITY_LABELS=true`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: workflows/media/workflow.go [hunk 1]
```diff
@@ -30,10 +30,12 @@ func (e *stepError) Unwrap() error { return e.err }
 // Each activity is invoked with the retry policy that fits its idempotency
 // profile: idempotent operations (Notify, Cleanup) use retryableMaxAttempts
 // so transient flakes do not fail the run; non-idempotent operations
-// (RecordRunMetrics, RecordInvalid, NotifyFailure) use defaultMaxAttempts
-// (single attempt) so retries cannot duplicate side effects, and the workflow
-// swallows their errors so a metrics or webhook issue does not fail an
-// otherwise-successful run.
+// (transcode, failure-webhook) use defaultMaxAttempts (single attempt).
+//
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply workflows/media/workflow.go:36]: Updated the comment to cover every single-attempt activity (Probe, DetectCrop, Transcode, NotifyFailure) rather than calling out only transcode + failure-webhook. Fixed in beb39ab.

File: workflows/media/activities_test.go [hunk 2]
```diff
@@ -9,42 +9,45 @@ import (
...
+	}
+
 	env := suite.NewTestActivityEnvironment()
+	env.RegisterActivity(a.Probe)
 	env.RegisterActivity(a.Notify)
 	env.RegisterActivity(a.Cleanup)
-	env.RegisterActivity(a.RecordRunMetrics)
-	env.RegisterActivity(a.RecordInvalid)
+	env.RegisterActivity(a.Transcode)
 	env.RegisterActivity(a.NotifyFailure)
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply workflows/media/activities_test.go:47]: Registered DetectCrop in newActivityEnv so it now genuinely matches the "all activities" docstring (Probe, DetectCrop, Transcode, Notify, Cleanup, NotifyFailure). Fixed in beb39ab.

File: workflows/media/activities.go [hunk 3]
```diff
@@ -148,13 +159,74 @@ func (a *Activities) Transcode(ctx context.Context, input MediaInput, probe step
...
+func (a *Activities) resolveHighCardinalityLabels(ctx context.Context, input MediaInput, library medialib.ArrLibrary) map[string]string {
+	info, err := library.GetInfo(ctx, input.FilePath)
+	if err != nil || info == nil {
+		if err != nil {
+			activity.GetLogger(ctx).Warn("media workflow metrics: GetInfo failed", "error", err)
+		}
+
+		activity.GetMetricsHandler(ctx).Counter(metricMetricsErrors).Inc(1)
+
+		return nil
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply workflows/media/activities.go]: Good catch — this is the same registration-conflict pattern as the movie-vs-show case, which I'd guarded against in the success branch but missed on the failure path. Confirmed by inspection: tally caches on `(name, sortedTagKeys)` and the second registration with a different key set silently becomes a `noopMetric`, so before this fix any worker that hit a `GetInfo` failure would have had its file-size histogram observations dropped from that point forward.

Fixed in beb39ab — `resolveHighCardinalityLabels` now always returns the full six-key map (id, title, year, series_title, season_number, episode_number); on failure every value is the empty string. The metrics-errors counter still increments. Added `TestResolveHighCardinalityLabels_StableKeySetAcrossOutcomes` (success / GetInfo-error / nil-info subtests) that asserts the returned key set is identical across all three outcomes — guards against this pattern silently re-appearing.

File: docs/metrics.md [hunk 4]
```diff
@@ -2,14 +2,12 @@
...
-```sh
-# Worker default; override only if needed
-METRICS_ADDR=:9090
-# Watcher default
-METRICS_ADDR=:9091
-```
+| Binary  | Default metrics address | Override env var |
+| ------- | ----------------------- | ---------------- |
+| Worker  | `:9090`                 | `METRICS_ADDR`   |
+| Watcher | `:9091`                 | `METRICS_ADDR`   |
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply docs/metrics.md:10]: this comment is garbage

File: workflows/media/activities.go [hunk 5]
```diff
@@ -148,13 +159,82 @@ func (a *Activities) Transcode(ctx context.Context, input MediaInput, probe step
...
+	scope := scopedHistograms(handler, tags)
+	scope.Histogram(metricSourceFileSizeBytes, fileSizeBuckets).RecordValue(float64(transcode.SourceFileSizeBytes))
+	scope.Histogram(metricDestFileSizeBytes, fileSizeBuckets).RecordValue(float64(transcode.DestFileSizeBytes))
+	scope.Histogram(metricTranscodeDurationSecs, durationBuckets).
+		RecordDuration(time.Duration(transcode.TranscodeDurationSeconds * float64(time.Second)))
+
+	if transcode.ArtworkFetchSkipped {
+		handler.Counter(metricArtworkFetchSkipped).Inc(1)
+	}
 }
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply workflows/media/activities.go:239]: Added `TestEmitTranscodeMetrics_FullTagSetAndArtworkCounter` in 14d7526, covering three subcases (no HC, artwork skipped, HC tags merged in). Asserts the three histograms (`source_file_size_bytes`, `destination_file_size_bytes`, `transcode_duration_seconds`) all carry the full transcode tag set with one sample, HC tags merge in correctly when the map is supplied, and the artwork-skipped counter only fires when the `TranscodeOutput` flag is set.

File: workflows/media/activities_test.go [hunk 6]
```diff
@@ -189,8 +166,242 @@ func TestNotifyFailure_SendsWebhookPayload(t *testing.T) {
...
+	require.NoError(t, err)
+
+	snap := scope.Snapshot()
+	wantTags := map[string]string{"media_type": "movie", "mapping_name": "downloads"}
+
+	counter := findCounter(t, snap, "media_workflow_invalid_files", wantTags)
+	assert.EqualValues(t, 1, counter.Value())
+
+	assert.Empty(t, findHistograms(snap, "media_workflow_audio_track_count"), "track histograms must not be emitted for invalid files")
+	assert.Empty(t, findHistograms(snap, "media_workflow_subtitle_track_count"))
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply workflows/media/activities_test.go]: Real catch — confirmed by inspection. The assertions were calling `findHistograms` for `media_workflow_audio_track_count` / `media_workflow_subtitle_track_count`, but those metrics are now Gauges, so the snapshot's `Histograms()` map never contained them in the first place. The asserts were vacuously true: even if the gauges had been (wrongly) emitted on the invalid-media path, the test would still have passed.

Fixed in 14d7526 — switched the two track-count assertions to `findGauges`, kept `findHistograms` for `source_duration_seconds`, and added an inline note about the trap so the next person to touch this doesn't fall into it.

File: workflows/media/config.go [hunk 7]
```diff
@@ -48,10 +44,10 @@ const (
 	defaultFinalizeTimeout = 10 * time.Minute
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply workflows/media/config.go:44]: Updated to list the current post-transcode activities (notify, cleanup, failure-webhook). Fixed in 14d7526.

File: docs/metrics.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (solidDoWant) [reply docs/metrics.md:10]: Acknowledging the owner dismissal — verified the tables in this file all start with single `|`. No change needed.

File: docs/metrics.md [hunk 8]
```diff
@@ -20,49 +18,47 @@ scrape_configs:
...
+| `media_workflow_audio_track_count`           | gauge     | tracks  | Audio track count from the most recent probe per label combination.                                                            |
+| `media_workflow_subtitle_track_count`        | gauge     | tracks  | Subtitle track count from the most recent probe per label combination.                                                         |
+| `media_workflow_invalid_files_total`         | counter   | —       | Files skipped because they could not be probed or contained no video stream.                                                   |
+| `media_workflow_artwork_fetch_skipped_total` | counter   | —       | Transcode runs where artwork fetch was attempted but yielded no embeddable image.                                              |
+| `media_workflow_metrics_errors_total`        | counter   | —       | Metric-tagging metadata lookups that failed (Radarr/Sonarr unreachable while resolving high-cardinality tags).                 |
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply docs/metrics.md]: Good catch on the doc/code mismatch. Investigated the GetInfo error semantics:

- `GetInfo` calls `GetMovieByFilePath` / `GetEpisodeByFilePath`, which use Radarr/Sonarr's `/parse` endpoint with the filename stem (`pkg/medialib/radarr/client.go:42-83`).
- Returns `medialib.ErrNotFound` when the parse endpoint can't match the filename to a known library item — semantically "filename Radarr/Sonarr's parser couldn't recognize", not a "library doesn't have this file yet" condition (the parse endpoint identifies pre-import files by name).
- Returns wrapped errors for backend issues (auth, network, 5xx).

Both are real "couldn't tag the metric" signals — `ErrNotFound` typically points at filename-naming problems an operator would still want surfaced (those rows just won't get HC tags). I went with broadening the docstring rather than narrowing the increment so we don't lose visibility into parse misses.

Updated in 14d7526:

> Radarr/Sonarr `GetInfo` lookups that did not return a result while resolving high-cardinality tags — either a backend error (unreachable, auth failure, etc.) or the library could not parse the filename to a known item.

File: workflows/media/workflow.go [hunk 9]
```diff
@@ -29,11 +29,15 @@ func (e *stepError) Unwrap() error { return e.err }
...
-// otherwise-successful run.
+// so transient flakes do not fail the run; everything else (Probe,
+// DetectCrop, Transcode, NotifyFailure) uses defaultMaxAttempts (single
+// attempt) because retrying would either duplicate non-idempotent side
+// effects or repeat expensive work that will not recover.
+//
+// Per-run application metrics are emitted by the activities themselves through
+// the SDK MetricsHandler. End-to-end workflow latency comes for free from the
+// SDK's temporal_workflow_endtoend_latency_seconds histogram (tagged with
+// workflow_type), so the workflow body does not emit a duration metric.
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply workflows/media/workflow.go:40]: Real catch — I missed the e2e suite when removing `media_workflow_total_duration_seconds`. Both `e2e/radarr_test.go` and `e2e/sonarr_test.go` polled on `media_workflow_total_duration_seconds_count` to wait for the workflow to finish; with that metric gone the `Eventually` loop would have timed out on every e2e run.

Fixed in abd5e66:

- **Wait condition:** switched to `media_workflow_transcode_duration_seconds_count` (still mapping_name-tagged, emitted at the end of the `Transcode` activity — just before `Notify`+`Cleanup`). The output-file/codec checks that run before this assertion already prove the workflow actually completed; the metrics block is now exercising the metrics pipeline rather than gating workflow completion on it.
- **SDK metric assertion added:** new `temporal_workflow_endtoend_latency_seconds_count{workflow_type="MediaWorkflow"} >= 1` check that proves the documented replacement metric is actually landing on `/metrics`. This metric only carries SDK tags (no `mapping_name`), so the filter is on `workflow_type` instead.
- Updated the now-stale "record_metrics task" comment and the metric-name example in `helpers_test.go`.

Built with `-tags=e2e` to confirm the suite still compiles. Actual e2e run still needs a Docker host before merge.

File: workflows/media/activities_test.go [hunk 10]
```diff
@@ -189,8 +166,372 @@ func TestNotifyFailure_SendsWebhookPayload(t *testing.T) {
...
+// subset of tags, ignoring SDK-injected worker tags
+// (activity_type, namespace, task_queue, workflow_type) that vary by env.
+func hasSubsetTags(actual, want map[string]string) bool {
+	for k, v := range want {
+		if actual[k] != v {
+			return false
+		}
+	}
+
+	return true
```
[the remark this answers is not in the document]
Comment (solidDoWant) [reply workflows/media/activities_test.go:431]: not sure if this is accurate and/or relevant to the current codebase

[the remark this answers is not in the document]
Comment (solidDoWant) [reply workflows/media/activities_test.go:431]: @solidDoWant your skepticism is well-founded — verified that no current test in this PR exercises the edge case. Every `wantTags` literal in `activities_test.go` uses non-empty values, and the HC-tag subtest in `TestEmitTranscodeMetrics_FullTagSetAndArtworkCounter` checks the empty-string `series_title` / `season_number` / `episode_number` via a direct `extraTagAssertion` on the returned tags map, not through `hasSubsetTags`. So this isn't a current bug.

That said, the helper's contract reads "actual contains every key/value pair in want" — naturally implying the key must exist. With the current implementation a future test that wanted to assert "this HC tag is present but empty" would silently match snapshots that simply omitted the key, which is a real foot-gun.

Made the defensive fix in acea3f9 — switched to the two-value map lookup form so absence and empty-value are properly distinguished. Updated the docstring to explain why. No behavior change for any current test (verified `make test` clean).

Easy to revert if you'd rather not carry the change.
